@TimeToBuildBob I see in the logs

```
2026-04-27T19:05:01.767401Z	Cloning repository...
2026-04-27T19:05:03.474968Z	From https://github.com/gptme/gptme
2026-04-27T19:05:03.475329Z	 * branch            f240c9d7ea15f56d92c4cb5dbc6228db4f75aea9 -> FETCH_HEAD
2026-04-27T19:05:03.47546Z	
2026-04-27T19:05:03.628894Z	HEAD is now at f240c9d fix(webui/voice): use type=text for voice server URL input
2026-04-27T19:05:03.629276Z	
2026-04-27T19:05:03.689775Z	
2026-04-27T19:05:03.690194Z	Using v2 root directory strategy
2026-04-27T19:05:03.705936Z	Success: Finished cloning repository files
2026-04-27T19:05:05.543376Z	Checking for configuration in a Wrangler configuration file (BETA)
2026-04-27T19:05:05.543797Z	
2026-04-27T19:05:06.678676Z	No Wrangler configuration file found. Continuing.
2026-04-27T19:05:06.870865Z	Detected the following tools from environment: poetry@2.1.3, pip@25.1.1, python@3.13.3, npm@10.9.2, bun@1.2.15, nodejs@22.16.0
2026-04-27T19:05:06.871738Z	Installing project dependencies: bun install --frozen-lockfile
2026-04-27T19:05:07.069669Z	bun install v1.2.15 (df017990)
2026-04-27T19:05:10.604202Z	
2026-04-27T19:05:10.605352Z	+ @eslint/js@9.39.4
2026-04-27T19:05:10.605713Z	+ @playwright/test@1.59.1
2026-04-27T19:05:10.605782Z	+ @tailwindcss/typography@0.5.19
2026-04-27T19:05:10.606123Z	+ @testing-library/jest-dom@6.9.1
2026-04-27T19:05:10.60624Z	+ @testing-library/react@16.3.2
2026-04-27T19:05:10.606332Z	+ @types/jest@29.5.14
2026-04-27T19:05:10.606452Z	+ @types/node@22.19.17
2026-04-27T19:05:10.606532Z	+ @types/react@18.3.28
2026-04-27T19:05:10.606586Z	+ @types/react-dom@18.3.7
2026-04-27T19:05:10.606705Z	+ @vitejs/plugin-react-swc@3.11.0
2026-04-27T19:05:10.606805Z	+ @vitest/coverage-v8@3.2.4
2026-04-27T19:05:10.606882Z	+ autoprefixer@10.5.0
2026-04-27T19:05:10.606979Z	+ eslint@9.39.4
2026-04-27T19:05:10.607049Z	+ eslint-plugin-prettier@5.5.5
2026-04-27T19:05:10.607121Z	+ eslint-plugin-react-hooks@5.2.0
2026-04-27T19:05:10.608706Z	+ eslint-plugin-react-refresh@0.4.26
2026-04-27T19:05:10.609013Z	+ globals@15.15.0
2026-04-27T19:05:10.60978Z	+ identity-obj-proxy@3.0.0
2026-04-27T19:05:10.609901Z	+ jest@29.7.0
2026-04-27T19:05:10.61Z	+ jest-environment-jsdom@29.7.0
2026-04-27T19:05:10.610129Z	+ jsdom@26.1.0
2026-04-27T19:05:10.6108Z	+ postcss@8.5.12
2026-04-27T19:05:10.611125Z	+ prettier@3.8.3
2026-04-27T19:05:10.611231Z	+ prettier-plugin-tailwindcss@0.5.14
2026-04-27T19:05:10.611323Z	+ tailwindcss@3.4.19
2026-04-27T19:05:10.61177Z	+ ts-jest@29.4.9
2026-04-27T19:05:10.612105Z	+ ts-node@10.9.2
2026-04-27T19:05:10.613054Z	+ typescript@5.9.3
2026-04-27T19:05:10.613283Z	+ typescript-eslint@8.59.0
2026-04-27T19:05:10.613481Z	+ vite@5.4.21
2026-04-27T19:05:10.613672Z	+ vitest@3.2.4
2026-04-27T19:05:10.614142Z	+ @hookform/resolvers@3.10.0
2026-04-27T19:05:10.614238Z	+ @icons-pack/react-simple-icons@13.13.0
2026-04-27T19:05:10.614543Z	+ @legendapp/state@3.0.0-beta.46
2026-04-27T19:05:10.614712Z	+ @radix-ui/react-accordion@1.2.12
2026-04-27T19:05:10.615233Z	+ @radix-ui/react-alert-dialog@1.1.15
2026-04-27T19:05:10.615427Z	+ @radix-ui/react-aspect-ratio@1.1.8
2026-04-27T19:05:10.615933Z	+ @radix-ui/react-avatar@1.1.11
2026-04-27T19:05:10.616239Z	+ @radix-ui/react-checkbox@1.3.3
2026-04-27T19:05:10.616742Z	+ @radix-ui/react-collapsible@1.1.12
2026-04-27T19:05:10.616988Z	+ @radix-ui/react-context-menu@2.2.16
2026-04-27T19:05:10.617238Z	+ @radix-ui/react-dialog@1.1.15
2026-04-27T19:05:10.617338Z	+ @radix-ui/react-dropdown-menu@2.1.16
2026-04-27T19:05:10.617528Z	+ @radix-ui/react-hover-card@1.1.15
2026-04-27T19:05:10.617903Z	+ @radix-ui/react-label@2.1.8
2026-04-27T19:05:10.618199Z	+ @radix-ui/react-menubar@1.1.16
2026-04-27T19:05:10.618468Z	+ @radix-ui/react-navigation-menu@1.2.14
2026-04-27T19:05:10.6187Z	+ @radix-ui/react-popover@1.1.15
2026-04-27T19:05:10.619207Z	+ @radix-ui/react-progress@1.1.8
2026-04-27T19:05:10.619652Z	+ @radix-ui/react-radio-group@1.3.8
2026-04-27T19:05:10.619745Z	+ @radix-ui/react-scroll-area@1.2.10
2026-04-27T19:05:10.619966Z	+ @radix-ui/react-select@2.2.6
2026-04-27T19:05:10.620315Z	+ @radix-ui/react-separator@1.1.8
2026-04-27T19:05:10.620401Z	+ @radix-ui/react-slider@1.3.6
2026-04-27T19:05:10.620458Z	+ @radix-ui/react-slot@1.2.4
2026-04-27T19:05:10.620595Z	+ @radix-ui/react-switch@1.2.6
2026-04-27T19:05:10.620678Z	+ @radix-ui/react-tabs@1.1.13
2026-04-27T19:05:10.620855Z	+ @radix-ui/react-toast@1.2.15
2026-04-27T19:05:10.621108Z	+ @radix-ui/react-toggle@1.1.10
2026-04-27T19:05:10.621342Z	+ @radix-ui/react-toggle-group@1.1.11
2026-04-27T19:05:10.62144Z	+ @radix-ui/react-tooltip@1.2.8
2026-04-27T19:05:10.621543Z	+ @tanstack/react-query@5.100.5
2026-04-27T19:05:10.621613Z	+ @tauri-apps/api@2.10.1
2026-04-27T19:05:10.621678Z	+ @tauri-apps/plugin-log@2.8.0
2026-04-27T19:05:10.621742Z	+ class-variance-authority@0.7.1
2026-04-27T19:05:10.62186Z	+ clsx@2.1.1
2026-04-27T19:05:10.621944Z	+ cmdk@1.1.1
2026-04-27T19:05:10.622008Z	+ date-fns@3.6.0
2026-04-27T19:05:10.622065Z	+ embla-carousel-react@8.6.0
2026-04-27T19:05:10.622232Z	+ highlight.js@11.11.1
2026-04-27T19:05:10.6224Z	+ input-otp@1.4.2
2026-04-27T19:05:10.622591Z	+ lovable-tagger@1.2.0
2026-04-27T19:05:10.622831Z	+ lucide-react@0.451.0
2026-04-27T19:05:10.62319Z	+ marked@14.1.4
2026-04-27T19:05:10.623289Z	+ marked-highlight@2.2.4
2026-04-27T19:05:10.623348Z	+ next-themes@0.3.0
2026-04-27T19:05:10.623393Z	+ react@18.3.1
2026-04-27T19:05:10.623425Z	+ react-day-picker@8.10.1
2026-04-27T19:05:10.623467Z	+ react-dom@18.3.1
2026-04-27T19:05:10.623502Z	+ react-hook-form@7.74.0
2026-04-27T19:05:10.623542Z	+ react-resizable-panels@2.1.9
2026-04-27T19:05:10.623574Z	+ react-router-dom@6.30.3
2026-04-27T19:05:10.623623Z	+ recharts@2.15.4
2026-04-27T19:05:10.623659Z	+ simple-icons@15.22.0
2026-04-27T19:05:10.623691Z	+ sonner@1.7.4
2026-04-27T19:05:10.623772Z	+ tailwind-merge@2.6.1
2026-04-27T19:05:10.623911Z	+ tailwindcss-animate@1.0.7
2026-04-27T19:05:10.623972Z	+ vaul@0.9.9
2026-04-27T19:05:10.624026Z	+ zod@3.25.76
2026-04-27T19:05:10.624097Z	
2026-04-27T19:05:10.624154Z	756 packages installed [3.58s]
2026-04-27T19:05:10.624206Z	
2026-04-27T19:05:10.624238Z	Blocked 1 postinstall. Run `bun pm untrusted` for details.
2026-04-27T19:05:10.635322Z	Executing user command: npm run build
2026-04-27T19:05:11.216455Z	
2026-04-27T19:05:11.216836Z	> gptme-webui@0.0.0 build
2026-04-27T19:05:11.216938Z	> tsc -p tsconfig.app.json --noEmit && vite build
2026-04-27T19:05:11.216994Z	
2026-04-27T19:05:24.19333Z	src/components/ProviderIcon.tsx(1,10): error TS2724: '"@icons-pack/react-simple-icons"' has no exported member named 'SiOpenai'. Did you mean 'SiOpenid'?
2026-04-27T19:05:24.193973Z	src/components/settings/McpConfiguration.tsx(42,27): error TS2345: Argument of type '{ env: { key: string; value: string; }[]; name?: string | undefined; enabled?: boolean | undefined; args?: string | undefined; command?: string | undefined; }' is not assignable to parameter of type '{ name: string; enabled: boolean; args: string; command: string; env?: { key: string; value: string; }[] | undefined; }'.
2026-04-27T19:05:24.194232Z	  Types of property 'name' are incompatible.
2026-04-27T19:05:24.194398Z	    Type 'string | undefined' is not assignable to type 'string'.
2026-04-27T19:05:24.194488Z	      Type 'undefined' is not assignable to type 'string'.
2026-04-27T19:05:24.194607Z	src/components/settings/McpConfiguration.tsx(72,25): error TS2345: Argument of type '{ env: { key: string; value: string; }[]; name?: string | undefined; enabled?: boolean | undefined; args?: string | undefined; command?: string | undefined; }' is not assignable to parameter of type '{ name: string; enabled: boolean; args: string; command: string; env?: { key: string; value: string; }[] | undefined; }'.
2026-04-27T19:05:24.194779Z	  Types of property 'name' are incompatible.
2026-04-27T19:05:24.194911Z	    Type 'string | undefined' is not assignable to type 'string'.
2026-04-27T19:05:24.195028Z	      Type 'undefined' is not assignable to type 'string'.
2026-04-27T19:05:24.296612Z	Failed: Error while executing user command. Exited with error code: 2
2026-04-27T19:05:24.303775Z	Failed: build command exited with code: 1
2026-04-27T19:05:24.953071Z	Failed: error occurred while running build command
```

Maybe an issue with it using bun? Open separate issue for that issue, it has never worked.

--- a/webui/public/pcm-recorder-worklet.js
+++ b/webui/public/pcm-recorder-worklet.js
@@ -1,0 +1,41 @@
+/**
+ * AudioWorklet processor: downsamples mic input to 16 kHz PCM16 LE mono.
+ *
+ * Loaded at runtime via audioContext.audioWorklet.addModule('/pcm-recorder-worklet.js').
+ * Must be plain JS (no ES module imports) — AudioWorklet scope has no module loader.
+ */
+class PCMRecorderProcessor extends AudioWorkletProcessor {
+  constructor() {
+    super();
+    // ratio is computed on first process() call when sampleRate global is available
+    this._ratio = null;
+    this._targetRate = 16000;
+  }
+
+  process(inputs) {
+    const input = inputs[0];
+    if (!input || !input[0]) return true;
+    const samples = input[0]; // Float32Array at AudioContext native rate
+
+    if (this._ratio === null) {
+      // sampleRate is a global in AudioWorklet scope
+      this._ratio = sampleRate / this._targetRate;
+    }
+
+    // Simple nearest-neighbor downsampling (acceptable quality for speech)
+    const outputLength = Math.floor(samples.length / this._ratio);
+    if (outputLength === 0) return true;
+
+    const output = new Int16Array(outputLength);
+    for (let i = 0; i < outputLength; i++) {
+      const srcIdx = Math.min(Math.round(i * this._ratio), samples.length - 1);
+      const s = Math.max(-1, Math.min(1, samples[srcIdx]));
+      output[i] = s < 0 ? Math.round(s * 32768) : Math.round(s * 32767);
+    }
+
+    this.port.postMessage(output.buffer, [output.buffer]);
+    return true;
+  }
+}
+
+registerProcessor('pcm-recorder-processor', PCMRecorderProcessor);

--- a/webui/src/audio/pcm-player.ts
+++ b/webui/src/audio/pcm-player.ts
@@ -1,0 +1,70 @@
+/**
+ * PCMPlayer: buffers and schedules raw PCM16 LE playback at a fixed sample rate.
+ *
+ * Incoming binary frames from the voice server are raw Int16 mono PCM (no container).
+ * We hand-build AudioBuffer objects (decodeAudioData is for container formats, not raw PCM)
+ * and schedule them end-to-end so consecutive utterances never gap.
+ */
+
+// Extend Window to cover webkit-prefixed AudioContext
+interface WindowWithWebkit extends Window {
+  webkitAudioContext?: typeof AudioContext;
+}
+
+export class PCMPlayer {
+  private ctx: AudioContext;
+  private sampleRate: number;
+  private nextStart: number;
+
+  constructor(sampleRate = 24000) {
+    const win = window as WindowWithWebkit;
+    const AudioCtx = window.AudioContext ?? win.webkitAudioContext;
+    if (!AudioCtx) throw new Error('AudioContext not supported');
+    this.ctx = new AudioCtx();
+    this.sampleRate = sampleRate;
+    this.nextStart = 0;
+  }
+
+  /** Queue a raw PCM16 LE ArrayBuffer for gapless playback. */
+  feed(buffer: ArrayBuffer): void {
+    if (buffer.byteLength === 0) return;
+
+    const int16 = new Int16Array(buffer);
+    const float32 = new Float32Array(int16.length);
+    for (let i = 0; i < int16.length; i++) {
+      float32[i] = int16[i] / 32768;
+    }
+
+    const audioBuffer = this.ctx.createBuffer(1, float32.length, this.sampleRate);
+    audioBuffer.copyToChannel(float32, 0);
+
+    const source = this.ctx.createBufferSource();
+    source.buffer = audioBuffer;
+    source.connect(this.ctx.destination);
+
+    const now = this.ctx.currentTime;
+    const when = Math.max(now + 0.01, this.nextStart); // 10 ms lookahead
+    source.start(when);
+    this.nextStart = when + audioBuffer.duration;
+  }
+
+  /** Resume suspended AudioContext (required after user gesture in some browsers). */
+  async resume(): Promise<void> {
+    if (this.ctx.state === 'suspended') {
+      await this.ctx.resume();
+    }
+  }
+
+  /** Reset scheduling cursor (call between utterances if needed). */
+  reset(): void {
+    this.nextStart = 0;
+  }
+
+  close(): void {
+    void this.ctx.close();
+  }
+
+  get audioContext(): AudioContext {
+    return this.ctx;
+  }
+}

--- a/webui/src/components/ChatInput.tsx
+++ b/webui/src/components/ChatInput.tsx
@@ -47,6 +47,8 @@ import { useModels } from '@/hooks/useModels';
 import { useAgents } from '@/hooks/useAgents';
 import { useFileAutocomplete } from '@/hooks/useFileAutocomplete';
 import { FileAutocomplete } from '@/components/FileAutocomplete';
+import { VoiceButton } from '@/components/VoiceButton';
+import { useSettings } from '@/contexts/SettingsContext';
 import {
   Select,
   SelectContent,
@@ -461,6 +463,7 @@ export const ChatInput: FC<Props> = ({
   onEditCancel,
 }) => {
   const { isConnected$, connectionConfig } = useApi();
+  const { settings } = useSettings();
   const sidebarSelectedWorkspace = use$(selectedWorkspace$);
   const sidebarSelectedAgent = use$(selectedAgent$);
 
@@ -1118,6 +1121,9 @@ export const ChatInput: FC<Props> = ({
                         )}
                     </div>
 
+                    {settings.voiceServerUrl && (
+                      <VoiceButton voiceServerUrl={settings.voiceServerUrl} />
+                    )}
                     <SubmitButton
                       isGenerating={isGenerating}
                       isDisabled={isDisabled}

--- a/webui/src/components/SettingsModal.tsx
+++ b/webui/src/components/SettingsModal.tsx
@@ -8,6 +8,7 @@ import {
   DialogTrigger,
 } from '@/components/ui/dialog';
 import { Button } from '@/components/ui/button';
+import { Input } from '@/components/ui/input';
 import { Label } from '@/components/ui/label';
 import { Switch } from '@/components/ui/switch';
 import { Separator } from '@/components/ui/separator';
@@ -195,6 +196,24 @@ export const SettingsModal = forwardRef<HTMLButtonElement, SettingsModalProps>(
                   id="chime-toggle"
                   checked={settings.chimeEnabled}
                   onCheckedChange={(checked) => updateSettings({ chimeEnabled: checked })}
+                />
+              </div>
+
+              <div className="space-y-2">
+                <Label htmlFor="voice-server-url" className="text-sm">
+                  Voice server URL
+                </Label>
+                <p className="text-xs text-muted-foreground">
+                  WebSocket URL of a running gptme-voice-server instance, e.g.{' '}
+                  <code className="rounded bg-muted px-1">ws://localhost:5700/voice</code>. Leave
+                  empty to hide the voice button.
+                </p>
+                <Input
+                  id="voice-server-url"
+                  type="url"
+                  placeholder="ws://localhost:5700/voice"
+                  value={settings.voiceServerUrl}
+                  onChange={(e) => updateSettings({ voiceServerUrl: e.target.value.trim() })}
                 />
               </div>
             </div>

--- a/webui/src/components/VoiceButton.tsx
+++ b/webui/src/components/VoiceButton.tsx
@@ -1,0 +1,79 @@
+import { Mic, MicOff, Loader2 } from 'lucide-react';
+import { Button } from '@/components/ui/button';
+import { Tooltip, TooltipContent, TooltipTrigger } from '@/components/ui/tooltip';
+import { useVoiceSession, type VoiceState } from '@/hooks/useVoiceSession';
+
+interface Props {
+  voiceServerUrl: string;
+}
+
+function stateLabel(s: VoiceState, error: string | null): string {
+  if (error) return `Voice error: ${error}`;
+  switch (s) {
+    case 'connecting':
+      return 'Connecting to voice server…';
+    case 'recording':
+      return 'Recording — click to stop';
+    case 'ended':
+      return 'Ending session…';
+    default:
+      return 'Start voice session';
+  }
+}
+
+export function VoiceButton({ voiceServerUrl }: Props) {
+  const { state, error, level, start, stop } = useVoiceSession(voiceServerUrl);
+
+  const isActive = state === 'recording' || state === 'connecting';
+  const isLoading = state === 'connecting';
+  const hasError = !!error;
+
+  const handleClick = () => {
+    if (isActive) stop();
+    else start();
+  };
+
+  // Mic level ring: scale from 1.0 to 1.3 based on audio level
+  const ringScale = 1 + level * 0.3;
+
+  return (
+    <Tooltip>
+      <TooltipTrigger asChild>
+        <Button
+          type="button"
+          variant="ghost"
+          size="icon"
+          onClick={handleClick}
+          className={[
+            'relative h-8 w-8 shrink-0 rounded-full transition-colors',
+            hasError ? 'text-destructive' : '',
+            state === 'recording' ? 'text-red-500 hover:text-red-600' : 'text-muted-foreground',
+          ].join(' ')}
+          aria-label={stateLabel(state, error)}
+          aria-pressed={isActive}
+        >
+          {/* Level ring behind the icon */}
+          {state === 'recording' && (
+            <span
+              className="absolute inset-0 rounded-full bg-red-500/20 transition-transform duration-75"
+              style={{ transform: `scale(${ringScale})` }}
+            />
+          )}
+
+          {isLoading ? (
+            <Loader2 className="h-4 w-4 animate-spin" />
+          ) : isActive ? (
+            <Mic className="relative h-4 w-4" />
+          ) : hasError ? (
+            <MicOff className="h-4 w-4" />
+          ) : (
+            <Mic className="h-4 w-4" />
+          )}
+        </Button>
+      </TooltipTrigger>
+      <TooltipContent side="top">
+        <p>{stateLabel(state, error)}</p>
+      </TooltipContent>
+    </Tooltip>
+  );
+}

--- a/webui/src/components/__tests__/ChatInput.test.tsx
+++ b/webui/src/components/__tests__/ChatInput.test.tsx
@@ -5,6 +5,22 @@ import { ChatInput } from '../ChatInput';
 
 const mockUploadFiles = jest.fn();
 
+jest.mock('@/contexts/SettingsContext', () => ({
+  useSettings: () => ({
+    settings: {
+      chimeEnabled: true,
+      blocksDefaultOpen: true,
+      showHiddenMessages: false,
+      showInitialSystem: false,
+      hasCompletedSetup: true,
+      welcomeBackground: '',
+      voiceServerUrl: '',
+    },
+    updateSettings: jest.fn(),
+    resetSettings: jest.fn(),
+  }),
+}));
+
 jest.mock('@/contexts/ApiContext', () => {
   const { observable } = jest.requireActual('@legendapp/state');
   return {

--- a/webui/src/contexts/SettingsContext.tsx
+++ b/webui/src/contexts/SettingsContext.tsx
@@ -8,6 +8,11 @@ export interface Settings {
   hasCompletedSetup: boolean;
   /** CSS background for the welcome/new-chat view (image URL or gradient) */
   welcomeBackground: string;
+  /**
+   * WebSocket URL for the gptme-voice-server /voice endpoint, e.g. ws://localhost:5700/voice.
+   * Leave empty to hide the VoiceButton.
+   */
+  voiceServerUrl: string;
 }
 
 interface SettingsContextType {
@@ -23,6 +28,7 @@ const defaultSettings: Settings = {
   showInitialSystem: false,
   hasCompletedSetup: false,
   welcomeBackground: '',
+  voiceServerUrl: '',
 };
 
 function loadSettingsFromStorage(): Settings {

--- a/webui/src/hooks/useVoiceSession.ts
+++ b/webui/src/hooks/useVoiceSession.ts
@@ -1,0 +1,160 @@
+import { useState, useRef, useCallback, useEffect } from 'react';
+import { PCMPlayer } from '@/audio/pcm-player';
+
+export type VoiceState = 'idle' | 'connecting' | 'recording' | 'ended';
+
+export interface UseVoiceSessionReturn {
+  state: VoiceState;
+  error: string | null;
+  /** Level 0–1 from mic analyser (0 when not recording). */
+  level: number;
+  start: () => void;
+  stop: () => void;
+  /** Flush the server's VAD input buffer (optional push-to-talk signal). */
+  commit: () => void;
+}
+
+interface Session {
+  ws: WebSocket;
+  player: PCMPlayer;
+  audioCtx: AudioContext;
+  stream: MediaStream;
+  rafId: number;
+}
+
+export function useVoiceSession(voiceServerUrl: string): UseVoiceSessionReturn {
+  const [state, setState] = useState<VoiceState>('idle');
+  const [error, setError] = useState<string | null>(null);
+  const [level, setLevel] = useState(0);
+
+  const sessionRef = useRef<Session | null>(null);
+
+  const cleanup = useCallback(() => {
+    const s = sessionRef.current;
+    if (!s) return;
+    sessionRef.current = null;
+
+    cancelAnimationFrame(s.rafId);
+    s.stream.getTracks().forEach((t) => t.stop());
+    s.ws.onopen = null;
+    s.ws.onmessage = null;
+    s.ws.onerror = null;
+    s.ws.onclose = null;
+    if (s.ws.readyState <= WebSocket.OPEN) s.ws.close();
+    s.player.close();
+    void s.audioCtx.close();
+    setLevel(0);
+  }, []);
+
+  const start = useCallback(() => {
+    if (!voiceServerUrl || sessionRef.current) return;
+
+    setError(null);
+    setState('connecting');
+
+    void (async () => {
+      try {
+        const stream = await navigator.mediaDevices.getUserMedia({ audio: true });
+        const ctx = new AudioContext();
+
+        await ctx.audioWorklet.addModule('/pcm-recorder-worklet.js');
+        const workletNode = new AudioWorkletNode(ctx, 'pcm-recorder-processor');
+
+        // Level meter
+        const analyser = ctx.createAnalyser();
+        analyser.fftSize = 256;
+
+        // Route mic → analyser + worklet; mute to speakers
+        const silentGain = ctx.createGain();
+        silentGain.gain.value = 0;
+        const micSource = ctx.createMediaStreamSource(stream);
+        micSource.connect(analyser);
+        micSource.connect(workletNode);
+        workletNode.connect(silentGain);
+        silentGain.connect(ctx.destination);
+
+        const player = new PCMPlayer(24000);
+        const ws = new WebSocket(voiceServerUrl);
+        ws.binaryType = 'arraybuffer';
+
+        // Stash session before async events can fire
+        const session: Session = { ws, player, audioCtx: ctx, stream, rafId: 0 };
+        sessionRef.current = session;
+
+        ws.onmessage = (evt) => {
+          if (typeof evt.data === 'string') {
+            try {
+              const msg = JSON.parse(evt.data) as { type: string };
+              if (msg.type === 'ready') {
+                void player.resume();
+                setState('recording');
+
+                // Wire worklet output → WebSocket
+                workletNode.port.onmessage = (e: MessageEvent<ArrayBuffer>) => {
+                  if (ws.readyState === WebSocket.OPEN) ws.send(e.data);
+                };
+
+                // Start level meter loop
+                const levelBuf = new Uint8Array(analyser.frequencyBinCount);
+                const tick = () => {
+                  analyser.getByteFrequencyData(levelBuf);
+                  const avg = levelBuf.reduce((sum, v) => sum + v, 0) / levelBuf.length;
+                  setLevel(Math.min(1, avg / 64));
+                  const id = requestAnimationFrame(tick);
+                  if (sessionRef.current === session) session.rafId = id;
+                };
+                session.rafId = requestAnimationFrame(tick);
+              } else if (msg.type === 'audio_end') {
+                player.reset();
+              }
+            } catch {
+              // non-JSON control frame — ignore
+            }
+          } else {
+            // Binary: raw PCM from the model
+            player.feed(evt.data as ArrayBuffer);
+          }
+        };
+
+        ws.onerror = () => {
+          setError('Voice connection error');
+          setState('ended');
+          cleanup();
+        };
+
+        ws.onclose = () => {
+          setState('ended');
+          cleanup();
+        };
+      } catch (err) {
+        setError(err instanceof Error ? err.message : 'Voice setup failed');
+        setState('ended');
+        cleanup();
+      }
+    })();
+  }, [voiceServerUrl, cleanup]);
+
+  const stop = useCallback(() => {
+    setState('ended');
+    cleanup();
+  }, [cleanup]);
+
+  const commit = useCallback(() => {
+    const ws = sessionRef.current?.ws;
+    if (ws?.readyState === WebSocket.OPEN) {
+      ws.send(JSON.stringify({ type: 'commit' }));
+    }
+  }, []);
+
+  // Auto-reset ended → idle so the button becomes clickable again
+  useEffect(() => {
+    if (state !== 'ended') return;
+    const t = setTimeout(() => setState('idle'), 1500);
+    return () => clearTimeout(t);
+  }, [state]);
+
+  // Cleanup on unmount
+  useEffect(() => () => cleanup(), [cleanup]);
+
+  return { state, error, level, start, stop, commit };
+}

--- a/webui/src/hooks/useVoiceSession.ts
+++ b/webui/src/hooks/useVoiceSession.ts
@@ -53,10 +53,11 @@ export function useVoiceSession(voiceServerUrl: string): UseVoiceSessionReturn {
     setState('connecting');
 
     void (async () => {
-      // Declared outside try so catch can release them if addModule fails
+      // Declared outside try so catch can release them if setup fails
       // before sessionRef is assigned (cleanup() would be a no-op otherwise).
       let stream: MediaStream | null = null;
       let ctx: AudioContext | null = null;
+      let player: PCMPlayer | null = null;
       try {
         stream = await navigator.mediaDevices.getUserMedia({ audio: true });
         ctx = new AudioContext();
@@ -80,12 +81,12 @@ export function useVoiceSession(voiceServerUrl: string): UseVoiceSessionReturn {
         workletNode.connect(silentGain);
         silentGain.connect(ctx.destination);
 
-        const player = new PCMPlayer(24000);
+        player = new PCMPlayer(24000);
         const ws = new WebSocket(voiceServerUrl);
         ws.binaryType = 'arraybuffer';
 
         // Stash session before async events can fire
-        const session: Session = { ws, player, audioCtx: ctx!, stream: stream!, rafId: 0 };
+        const session: Session = { ws, player: player!, audioCtx: ctx!, stream: stream!, rafId: 0 };
         sessionRef.current = session;
 
         ws.onmessage = (evt) => {
@@ -93,7 +94,7 @@ export function useVoiceSession(voiceServerUrl: string): UseVoiceSessionReturn {
             try {
               const msg = JSON.parse(evt.data) as { type: string };
               if (msg.type === 'ready') {
-                void player.resume();
+                void session.player.resume();
                 setState('recording');
 
                 // Wire worklet output → WebSocket
@@ -112,14 +113,14 @@ export function useVoiceSession(voiceServerUrl: string): UseVoiceSessionReturn {
                 };
                 session.rafId = requestAnimationFrame(tick);
               } else if (msg.type === 'audio_end') {
-                player.reset();
+                session.player.reset();
               }
             } catch {
               // non-JSON control frame — ignore
             }
           } else {
             // Binary: raw PCM from the model
-            player.feed(evt.data as ArrayBuffer);
+            session.player.feed(evt.data as ArrayBuffer);
           }
         };
 
@@ -135,9 +136,10 @@ export function useVoiceSession(voiceServerUrl: string): UseVoiceSessionReturn {
         };
       } catch (err) {
         // If setup failed before sessionRef was assigned, cleanup() is a no-op;
-        // release mic and recording context explicitly to prevent leaks.
+        // release mic, recording context, and player explicitly to prevent leaks.
         if (!sessionRef.current) {
           stream?.getTracks().forEach((t) => t.stop());
+          player?.close();
           void ctx?.close();
         }
         setError(err instanceof Error ? err.message : 'Voice setup failed');

